### PR TITLE
Update Angular to version 9

### PR DESCRIPTION
**Pull request by Nateowami**
**This PR is ready for review. It is only a draft because we don't want to merge it until after the migration on live is complete.**

This PR update Angular from version 8.2 to version 9.1. In the process of doing this, it was necessary to also update a number of other dependencies.

- I've made most of the changes one commit at a time in order to make it easier to follow the changes. I strongly encourage reviewing one commit at a time. Some of the early commits were generated by the Angular CLI as it upgraded packages and ran migrations (one commit per migration).
- For the earlier commits, I was following the official Angular upgrade instructions. To see the steps I was following, go to https://update.angular.io/#8.2:9.1l3 and tick "I use Angular Material".
- TypeScript was upgraded because Angular 9 supports newer versions.
- The version of @angular-mdc/web that we were using didn't work with Angular 9, so it had to be upgraded as well. A large part of the changes are fixing issues that arose due to this update.
- Prettier was upgraded because the version we were using didn't understand some newer TypeScript syntax that I happened to use in a fix.
- ng-bullet was removed because in Angular 9 it was completely broken, is no longer maintained, and appears to be unnecessary in Angular 9. A single run of the tests on master took 1 min 1.1 sec, while a run of the tests on this branch took 1 min 10.6 sec. With only two tests, and my CPU not at rest in either case, I consider the difference to be statistically insignificant.

Production build file sizes shrunk somewhat. Before:
```
chunk {0} runtime.458556a34b891ea32398.js, runtime.458556a34b891ea32398.js.map (runtime) 1.51 kB [entry] [rendered]
chunk {1} main.68264eeab34e55dbffcf.js, main.68264eeab34e55dbffcf.js.map (main) 2.9 MB [initial] [rendered]
chunk {2} polyfills.8523a15085cf86602c55.js, polyfills.8523a15085cf86602c55.js.map (polyfills) 66.5 kB [initial] [rendered]
chunk {3} polyfills-es5.4d0920dbfc91f86916b0.js, polyfills-es5.4d0920dbfc91f86916b0.js.map (polyfills-es5) 149 kB [initial] [rendered]
chunk {4} styles.76aa6016ea145cb990c1.css, styles.76aa6016ea145cb990c1.css.map (styles) 365 kB [initial] [rendered]
```
After:
```
chunk {0} runtime-es2015.0dae8cbc97194c7caed4.js, runtime-es2015.0dae8cbc97194c7caed4.js.map (runtime) 1.51 kB [entry] [rendered]
chunk {0} runtime-es5.0dae8cbc97194c7caed4.js, runtime-es5.0dae8cbc97194c7caed4.js.map (runtime) 1.51 kB [entry] [rendered]
chunk {2} polyfills-es2015.9097e8d5e7479e4a6ce2.js, polyfills-es2015.9097e8d5e7479e4a6ce2.js.map (polyfills) 59.8 kB [initial] [rendered]
chunk {3} polyfills-es5.1a76f58b0874c4e83578.js, polyfills-es5.1a76f58b0874c4e83578.js.map (polyfills-es5) 153 kB [initial] [rendered]
chunk {1} main-es2015.cff055263554e86ab979.js, main-es2015.cff055263554e86ab979.js.map (main) 2.22 MB [initial] [rendered]
chunk {1} main-es5.cff055263554e86ab979.js, main-es5.cff055263554e86ab979.js.map (main) 2.46 MB [initial] [rendered]
chunk {4} styles.197bc122be9c642649fb.css, styles.197bc122be9c642649fb.css.map (styles) 336 kB [initial] [rendered]
```

---
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/794)

### CODE REVIEW

**Review by Nateowami — COMMENTED** [review]
Note: The TeamCity build failure is not caused by failing tests. It's caused by Prettier changing its formatting standards (this PR upgrades Prettier) and the code coverage files not being found (not sure why). I'll be looking into this. I don't think it needs to hold up the review, since it's not caused by bugs or test failures.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/794#-:-MGMDSNw1BBMsM7EHsDR:bn0p8ae)* status: 0 of 75 files reviewed, all discussions resolved

File: src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.html
Comment (marksvc) [inline src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.html:22]: Seems odd to put a menu in a select. Ok if works well.

**Review by marksvc — COMMENTED** [review]
Thanks for all your work on this.
I did not carefully consider the more meaningful-looking CSS style changes.
I did not try the change in my browser.
Thank you for splitting the work into different commits with explanations about why things were being done.
I approve, minus comments.

File: src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.html
Comment (marksvc) [reply src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.html:22]: Done

File: src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.spec.ts
Comment (marksvc) [reply src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.spec.ts:229]: Done

File: src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.spec.ts
Comment (marksvc) [reply src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.spec.ts:620]: So let's use .toBe(true) and .toBeUndefined().

File: src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.spec.ts
Comment (marksvc) [reply src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.spec.ts:1618]: So let's use [elevated]

File: src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking-answers/checking-answers.component.html
Comment (marksvc) [reply src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking-answers/checking-answers.component.html:64]: Done

File: src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.spec.ts
Comment (marksvc) [reply src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.spec.ts:200]: (setting the value was not firing a needed event. Fixed by using .next)
Done. Tho consider `set confidenceThreshold`, and un-publicing `confidenceThreshold$`.

File: src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.html
Comment (Nateowami) [reply src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.html:22]: I think this was addressed in our discussion; that's just the way it's done now (no longer using native select).

File: src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.spec.ts
Comment (Nateowami) [reply src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.spec.ts:168]: Also discussed. `componentInstance` is probably `any`, which means TypeScript will let us do whatever we want with it.

File: src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.spec.ts
Comment (Nateowami) [reply src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.spec.ts:229]: Indeed. But as discussed, the old way is no longer possible because it no longer uses a native select.

You did mention though that you think elsewhere in the tests we click the menu, find the menu item we're looking for, and click it. I searched our codebase and found only two other instances like this, and in both we used this method. So for now we should probably leave it as-is, but an improvement could be made. Or perhaps there is a place where we do it differently and you may be able to locate it.

File: src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.spec.ts
Comment (Nateowami) [reply src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.spec.ts]: Thanks. I think you're right. I'll do a global find and replace.

File: src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.spec.ts
Comment (Nateowami) [reply src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.spec.ts]: Thanks. I'm making it slightly more specific.

File: .prettierrc.yml
Comment (Nateowami) [reply .prettierrc.yml:2]: I concur, but I think it should be discussed at least briefly with other developers first.

File: src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking-answers/checking-answers.component.html
Comment (Nateowami) [reply src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking-answers/checking-answers.component.html:64]: I think we discussed and resolved this.

File: src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.spec.ts
Comment (Nateowami) [reply src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.spec.ts]: Thanks. Done.

File: src/SIL.XForge.Scripture/ClientApp/src/app/checking/question-dialog/question-dialog.service.ts
Comment (Nateowami) [reply src/SIL.XForge.Scripture/ClientApp/src/app/checking/question-dialog/question-dialog.service.ts:31]: Glad it was clear.

**Comment by Nateowami** [comment]
Sorry, I see you made a number of comment a couple hours before I responded. All my comments were written without having seen your responses.

**Review by nigel-wells — COMMENTED** [review]
Great work @nateowami - besides making it all work there are some good tidy ups you've done as well. I've only made a couple of comments and not raised anything that you and Mark are already in discussions over.

> Reviewed 91 of 91 files at r1.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/794#-:-MHJyOVR5QOWW-Yz33l6:b5apjbg)* status: all files reviewed, 5 unresolved discussions (waiting on @Nateowami)

___
*[src/SIL.XForge.Scripture/ClientApp/src/app/app.component.ts, line 86 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/794#-MHJtbCaCwbiX61c494x:-MHJtbCaCwbiX61c494y:b-rf4c16) ([raw file](https://github.com/sillsdev/web-xforge/blob/361056885f5acf91e2399c5c5ae69721855d35b2/src/SIL.XForge.Scripture/ClientApp/src/app/app.component.ts#L86)):*
> ```
>  pwaService: PwaService,
>  iconRegistry: MdcIconRegistry,
>  sanitizer: DomSanitizer
> ```
Any reason to make these all public?
___
*[src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.html, line 27 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/794#-MHJCGXX54EjJn3UnGQw:-MHJCGXX54EjJn3UnGQx:bvsjdqw) ([raw file](https://github.com/sillsdev/web-xforge/blob/361056885f5acf91e2399c5c5ae69721855d35b2/src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.html#L27)):*
> ```HTML
>                   [projectUserConfigDoc]="projectUserConfigDoc"
>                   [project]="projectDoc?.data"
>                   [questionDocs]="questionDocs"
> ```

Why did these need to move?
___
*[src/SIL.XForge.Scripture/ClientApp/src/app/scripture-chooser-dialog/scripture-chooser-dialog.component.spec.ts, line 1 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/794#-MHJh-lj3e6PsAlkwTW3:-MHJh-lk1A9OGm8j2q0V:b-jj2kti) ([raw file](https://github.com/sillsdev/web-xforge/blob/361056885f5acf91e2399c5c5ae69721855d35b2/src/SIL.XForge.Scripture/ClientApp/src/app/scripture-chooser-dialog/scripture-chooser-dialog.component.spec.ts#L1)):*
> ```XML
> import { MDC_DIALOG_DATA, MdcDialog, MdcDialogConfig, MdcDialogModule, MdcDialogRef } from '@angular-mdc/web/dialog';
> ```

`MDC_DIALOG_DATA` also needs to be removed from import
___
*[src/SIL.XForge.Scripture/ClientApp/src/app/shared/chapter-nav/chapter-nav.component.scss, line 15 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/794#-MHJw0J1A6vygi4Lys8d:-MHJw0J1A6vygi4Lys8e:blowwwu) ([raw file](https://github.com/sillsdev/web-xforge/blob/361056885f5acf91e2399c5c5ae69721855d35b2/src/SIL.XForge.Scripture/ClientApp/src/app/shared/chapter-nav/chapter-nav.component.scss#L15)):*
> ```SCSS
> }
> 
> // Unfortunately, using Sass mixins to style the select was not successful at the current version of @angular-mdc/web
> ```

What was the reason out of interest? Did the import not return the expected structure that mdc web no outputs?
___
*[src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.scss, line 11 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/794#-MHJwQwc7UxUuJxDDV74:-MHJwQwc7UxUuJxDDV75:bteci68) ([raw file](https://github.com/sillsdev/web-xforge/blob/361056885f5acf91e2399c5c5ae69721855d35b2/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.scss#L11)):*
> ```SCSS
> mdc-dialog-content.content-padding {
>   padding-top: 15px;
>   overflow: visible; // prevent mdc-select menu from causing scroll bar
> ```

I've not tested this specifically with my project but is there still room on mobile without the scrollbar?

**Review by ddaspit — COMMENTED** [review]
Everything looks good to me.

> Reviewed 91 of 91 files at r1.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/794#-:-MHKkB4M-ij5drkucSHp:b6edp34)* status: all files reviewed, 5 unresolved discussions (waiting on @Nateowami)

**Review by Nateowami — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/794)* status: all files reviewed, 5 unresolved discussions (waiting on @nigel-wells)

___
*[src/SIL.XForge.Scripture/ClientApp/src/app/app.component.ts, line 86 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/794#-MHJtbCaCwbiX61c494x:-MHLoU4JD3JWPXrrSMQk:bwi6nuy) ([raw file](https://github.com/sillsdev/web-xforge/blob/361056885f5acf91e2399c5c5ae69721855d35b2/src/SIL.XForge.Scripture/ClientApp/src/app/app.component.ts#L86)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

> ```
>  pwaService: PwaService,
>  iconRegistry: MdcIconRegistry,
>  sanitizer: DomSanitizer
> ```
Any reason to make these all public?
</blockquote></details>

They're not public. They're not even properties. They're just arguments to the constructor. Marking them private or public causes them to become properties of the class.
___
*[src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.html, line 27 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/794#-MHJCGXX54EjJn3UnGQw:-MHLoweqEhQYlHMFSjlZ:b-k9fmz) ([raw file](https://github.com/sillsdev/web-xforge/blob/361056885f5acf91e2399c5c5ae69721855d35b2/src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.html#L27)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

Why did these need to move?
</blockquote></details>

To fix failing tests. The `app-checking-questions` component expects properties to be set in a particular order (For example, I think the setter for `questionDocs` assumes the `project` property is already set). How the order in which they were set previously happened to be right, I don't know. Really a user of a component shouldn't have to worry about the order. My conjecture is that the order in which Angular sets properties changed from one deterministic order to another (or the order in which some of our code is run was changed, possibly due to certain parts being run immediately vs. in a new task in the queue).

I also don't know why it was only a problem in the tests. A more robust solution would be for `questionDocs` setter to check whether the `project` property has been set, and if not, schedule its work to be done once it is set.
___
*[src/SIL.XForge.Scripture/ClientApp/src/app/scripture-chooser-dialog/scripture-chooser-dialog.component.spec.ts, line 1 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/794#-MHJh-lj3e6PsAlkwTW3:-MHLqsocEDlNkIPHXDIo:b-i21bbo) ([raw file](https://github.com/sillsdev/web-xforge/blob/361056885f5acf91e2399c5c5ae69721855d35b2/src/SIL.XForge.Scripture/ClientApp/src/app/scripture-chooser-dialog/scripture-chooser-dialog.component.spec.ts#L1)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

`MDC_DIALOG_DATA` also needs to be removed from import
</blockquote></details>

Thanks. Done.
___
*[src/SIL.XForge.Scripture/ClientApp/src/app/shared/chapter-nav/chapter-nav.component.scss, line 15 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/794#-MHJw0J1A6vygi4Lys8d:-MHMND-H24tX2CKW5i0i:b410sj2) ([raw file](https://github.com/sillsdev/web-xforge/blob/361056885f5acf91e2399c5c5ae69721855d35b2/src/SIL.XForge.Scripture/ClientApp/src/app/shared/chapter-nav/chapter-nav.component.scss#L15)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

What was the reason out of interest? Did the import not return the expected structure that mdc web no outputs?
</blockquote></details>

I can't remember, but I do remember that I tried a lot of mixins and nothing I tried was successful. Finally I decided to give up and just style it manually.
___
*[src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.scss, line 11 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/794#-MHJwQwc7UxUuJxDDV74:-MHMFaO74cZe_8OsI2NV:b-pm6xj1) ([raw file](https://github.com/sillsdev/web-xforge/blob/361056885f5acf91e2399c5c5ae69721855d35b2/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/suggestions-settings-dialog.component.scss#L11)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

I've not tested this specifically with my project but is there still room on mobile without the scrollbar?
</blockquote></details>

Yes. It just changes which element scrolls.

Without `overflow: visible` | With `overflow: visible`
----------------------------|-------------------------
 |

File: src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.spec.ts
Comment (marksvc) [reply src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.spec.ts]: Done.

**Comment by marksvc** [comment]
I approve, once other reviewers' concerns are satisfied.

**Review by nigel-wells — COMMENTED** [review]
All good with me.

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/794#-:-MHNd4ZzBgJ1sGPt9r76:bdeul4r)* status: all files reviewed, 2 unresolved discussions (waiting on @Nateowami)

___
*[src/SIL.XForge.Scripture/ClientApp/src/app/app.component.ts, line 86 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/794#-MHJtbCaCwbiX61c494x:-MHNcFSm6Gnmkhhv3Oa0:b6pq4b5) ([raw file](https://github.com/sillsdev/web-xforge/blob/361056885f5acf91e2399c5c5ae69721855d35b2/src/SIL.XForge.Scripture/ClientApp/src/app/app.component.ts#L86)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

They're not public. They're not even properties. They're just arguments to the constructor. Marking them private or public causes them to become properties of the class.
</blockquote></details>

#facepalm. In my defence Reviewable didn't show me enough context (without me expanding) to realise I was in a constructor.
___
*[src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.html, line 27 at r1](https://reviewable.io/reviews/sillsdev/web-xforge/794#-MHJCGXX54EjJn3UnGQw:-MHNd--b7khikzc3CDoY:b-f3jj89) ([raw file](https://github.com/sillsdev/web-xforge/blob/361056885f5acf91e2399c5c5ae69721855d35b2/src/SIL.XForge.Scripture/ClientApp/src/app/checking/checking/checking.component.html#L27)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

To fix failing tests. The `app-checking-questions` component expects properties to be set in a particular order (For example, I think the setter for `questionDocs` assumes the `project` property is already set). How the order in which they were set previously happened to be right, I don't know. Really a user of a component shouldn't have to worry about the order. My conjecture is that the order in which Angular sets properties changed from one deterministic order to another (or the order in which some of our code is run was changed, possibly due to certain parts being run immediately vs. in a new task in the queue).

I also don't know why it was only a problem in the tests. A more robust solution would be for `questionDocs` setter to check whether the `project` property has been set, and if not, schedule its work to be done once it is set.
</blockquote></details>

Weird. Bet that was a fun one to try and resolve!

**Review by Nateowami — COMMENTED** [review]
@marksvc @nigel-wells This is ready for re-review. I've squashed and moved commits around, and rebased on master. There may be a few other changes as well.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/794#-:-MHXq6T83tu7pIn948sQ:b667im1)* status: 85 of 93 files reviewed, 2 unresolved discussions (waiting on @ddaspit and @nigel-wells)

**Review by nigel-wells — APPROVED** [review]
Looks fine to me

> Reviewed 8 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/794#-:-MHia81c9c8KRZIic6T4:b1hxdoo)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**
